File: ios/Extopy/Features/Users/ProfileView.swift
//
//  ProfileView.swift
//  Extopy
//
//  Created by Nathan Fallet on 13/01/2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import shared
import SwiftUI
import Kingfisher
import KMPObservableViewModelSwiftUI
import KMPNativeCoroutinesAsync

struct ProfileView: View {
    
    @StateViewModel var viewModel: ProfileViewModel
    
    let viewedBy: CommonsUser
    
    var body: some View {
        ZStack {
            ScrollView {
                LazyVStack(spacing: 8) {
                    if let user = viewModel.user {
                        NavigationLink(destination: ProfileView(
                            viewModel: KoinApplication.shared.koin.profileViewModel(id: user.id),
                            viewedBy: viewedBy
                        )) {
                            UserCard(
                                user: user,
                                viewedBy: viewedBy,
                                onPostsClicked: { _ in },
                                onFollowersClicked: { _ in },
                                onFollowingClicked: { _ in },
                                onEditClicked: { _ in },
                                onFollowClicked: { _ in
                                    Task {
                                        try await asyncFunction(for: viewModel.onFollowClicked())
                                    }
                                },
                                onSettingsClicked: { _ in },
                                onDirectMessageClicked: { _ in }
                            )
                        }
                    }
                    ForEach(viewModel.posts ?? [], id: \.namespacedId) { post in
                        NavigationLink(destination: PostView(
                            viewModel: KoinApplication.shared.koin.postViewModel(id: post.id),
                            viewedBy: viewedBy
                        )) {
                            PostCard(
                                post: post,
                                viewedBy: viewedBy,
                                onLikeClicked: { post in
                                    Task {
                                        try await asyncFunction(for: viewModel.onLikeClicked(post: post))
                                    }
                                },
                                onRepostClicked: { _ in },
                                onReplyClicked: { _ in }
                            )
                            .onAppear {
                                viewModel.loadMoreIfNeeded(postId: post.id)
                            }
                        }
                    }
                }
                .padding()
            }
        }
        .navigationTitle("timeline_user_title")
        .refreshable {
            Task {
                try await asyncFunction(for: viewModel.fetchUser())
            }
        }
        .onAppear {
            Task {
                try await asyncFunction(for: viewModel.fetchUser())
            }
        }
    }
    
}

File: shared/src/commonMain/kotlin/com/extopy/ui/screens/users/ProfileView.kt
package com.extopy.ui.screens.users

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewModelScope
import com.extopy.models.navigation.Route
import com.extopy.models.users.User
import com.extopy.ui.components.posts.PostCard
import com.extopy.ui.components.users.UserCard
import com.extopy.viewmodels.users.ProfileViewModel
import dev.kaccelero.models.UUID
import extopy_app.shared.generated.resources.Res
import extopy_app.shared.generated.resources.timeline_user_title
import kotlinx.coroutines.launch
import org.jetbrains.compose.resources.stringResource
import org.koin.compose.viewmodel.koinViewModel
import org.koin.core.parameter.parametersOf

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ProfileView(
    id: UUID,
    viewedBy: User, // TODO: Manage on view model
    navigate: (Route) -> Unit,
    modifier: Modifier = Modifier,
) {

    val viewModel = koinViewModel<ProfileViewModel>(
        parameters = { parametersOf(id) }
    )

    LaunchedEffect(id) {
        viewModel.fetchUser()
    }

    val user by viewModel.user.collectAsState()
    val posts by viewModel.posts.collectAsState()

    Column(
        modifier
            .fillMaxSize()
            .background(color = MaterialTheme.colorScheme.background),
    ) {
        CenterAlignedTopAppBar(
            title = {
                Text(stringResource(Res.string.timeline_user_title))
            },
        )
        LazyColumn(
            verticalArrangement = Arrangement.spacedBy(8.dp),
            contentPadding = PaddingValues(16.dp, 16.dp, 16.dp, (16 + 88).dp),
        ) {
            item {
                user?.let {
                    UserCard(
                        user = it,
                        viewedBy = viewedBy,
                        navigate = navigate,
                        onPostsClicked = {},
                        onFollowersClicked = { user ->
                            //navigate("timelines/users/${user.id}/followers")
                        },
                        onFollowingClicked = { user ->
                            //navigate("timelines/users/${user.id}/following")
                        },
                        onEditClicked = {

                        },
                        onFollowClicked = {
                            viewModel.viewModelScope.launch {
                                viewModel.onFollowClicked()
                            }
                        },
                        onSettingsClicked = {

                        },
                        onDirectMessageClicked = {

                        }
                    )
                }
            }
            items(posts ?: listOf()) {
                PostCard(
                    post = it,
                    navigate = navigate,
                    onLikeClicked = { post ->
                        viewModel.viewModelScope.launch {
                            viewModel.onLikeClicked(post)
                        }
                    },
                    onRepostClicked = { post ->
                        navigate(Route.TimelineCompose(repostOfId = post.id.toString()))
                    },
                    onReplyClicked = { post ->
                        navigate(Route.TimelineCompose(repliedToId = post.id.toString()))
                    }
                )
                viewModel.loadMoreIfNeeded(it.id)
            }
            item {
                Spacer(modifier = Modifier.height(12.dp))
            }
        }
    }

}
